### PR TITLE
Polar flow import only

### PR DESCRIPTION
**Pull request by ssj2son-gohan-ne**
Basic function acquire tcx, fit, and gpx file
tcx-file is imported and afterwards all additional available excerise informations added

prerequisitory:
Polar Flow Acceslink Account - is for free
https://www.polar.com/accesslink-api/#how-to-get-started

### CODE REVIEW

**Comment by liversedge** [comment]
changes to yacc grammar files are breaking the builds. would suggest using an older version of bison on your build to avoid this.
we have not updated the build scripts due to version compatibility cross platform, so you must use an older version.

**Comment by ssj2son-gohan-ne** [comment]
okay.

Will try :)

**Comment by amtriathlon** [comment]
I think this is a mess, there is no reason for this PR to change Python embedding, or Bison parsers, or .gitignore, etc.

No change on unrelated files will be accepted, not even commented out ones or whitespace.

File: src/Cloud/PolarFlow.cpp [hunk 1]
```diff
@@ -49,105 +74,269 @@
...
+    if (containsSeconds == false) {
+        duration_str.prepend("S0.000");
+    }
+    printd("isoDateToSeconds - duration_String ISO after eventually prepend: %s\n", duration_str.toStdString().c_str());
+
+    // check contained letters H, M, S ; Hours, Minutes ? true / false
+    QString duration_strtime;
+
+    // Hours, Minutes and Seconds are provided in String
+    if (containsHours == true and containsMinutes == true and containsSeconds == true) {
```
Comment (amtriathlon) [inline src/Cloud/PolarFlow.cpp:296]: and is not a valid c++ operator, please use &&

File: src/Cloud/PolarFlow.cpp [hunk 2]
```diff
@@ -49,105 +74,269 @@
...
+        duration_strtime = duration_str.replace("PT", "").replace("H", ":").replace("M", ":").replace("S", ""); // PT4H51M38.512S->4:51:38.512
+    };
+    // Minutes are Zero - Hours und Seconds are provided in String
+    if (containsHours == true and containsMinutes == false and containsSeconds == true) {
```
Comment (amtriathlon) [inline src/Cloud/PolarFlow.cpp:300]: and is not a valid c++ operator, please use &&

File: src/Cloud/PolarFlow.cpp [hunk 3]
```diff
@@ -49,105 +74,269 @@
...
+        duration_strtime = duration_str.replace(QString("H"), QString("H0M"));
+        duration_strtime = duration_str.replace("PT", "").replace("H", ":").replace("M", ":").replace("S", ""); // PT43H8.512S->PT43H0M8.512S->4:00:38.512
+    };
+    // Hours are Zero - Minutes and Seconds are provided in String
+    if (containsHours == false and containsMinutes == true and containsSeconds == true) {
```
Comment (amtriathlon) [inline src/Cloud/PolarFlow.cpp:305]: and is not a valid c++ operator, please use &&

File: src/Cloud/PolarFlow.cpp [hunk 4]
```diff
@@ -49,105 +74,269 @@
...
+        duration_strtime = duration_str.replace("PT", "").prepend("0:").replace("M", ":").replace("S", ""); // 51M38.512S->0:51:38.512
+    };
+    // Hours and Minutes are Zero - Seconds are provided in String
+    if (containsHours == false and containsMinutes == false and containsSeconds == true) {
```
Comment (amtriathlon) [inline src/Cloud/PolarFlow.cpp:309]: and is not a valid c++ operator, please use &&

**Comment by amtriathlon** [comment]
Build is failing due to the use of and instead of && logical operator.

File: src/Core/DataFilter.y [hunk 5]
```diff
@@ -49,6 +49,14 @@ extern Leaf *DataFilterroot; // root node for parsed statement
...
 %}
 
+
+//#ifdef GC_WITH_BISON_EQGT_THREEPOINTSEVEN
+// Fix building with bison 3.7 #4119
+// generated by the scanner
+//   %define api.header.include {"DataFilter_yacc.h"}
+//#endif
+
+
```
Comment (amtriathlon) [inline src/Core/DataFilter.y:59]: Please, don't change unrelated files

File: src/Core/RideDB.y [hunk 6]
```diff
@@ -41,6 +41,12 @@ void RideDBerror(void*jc, const char *error) // used by parser aka yyerror()
 
 %}
 
+//#ifdef GC_WITH_BISON_EQGT_THREEPOINTSEVEN
+// Fix building with bison 3.7 #4119
+// generated by the scanner
+//   %define api.header.include {"RideDB_yacc.h"}
+//#endif
+
```
Comment (amtriathlon) [inline src/Core/RideDB.y:49]: Please, don't change unrelated files

File: src/FileIO/JsonRideFile.y [hunk 7]
```diff
@@ -106,6 +106,13 @@ static QString protect(const QString string)
...
 %}
 
+//#ifdef GC_WITH_BISON_EQGT_THREEPOINTSEVEN
+// Fix building with bison 3.7 #4119
+// generated by the scanner
+//   %define api.header.include {"JsonRideFile_yacc.h"}
+//#endif
+
+
 %pure-parser
```
Comment (amtriathlon) [inline src/FileIO/JsonRideFile.y:116]: Please, don't change unrelated files

File: src/src.pro [hunk 8]
```diff
@@ -705,7 +705,7 @@ HEADERS += Cloud/CalendarDownload.h Cloud/CloudService.h \
            Cloud/LocalFileStore.h Cloud/OAuthDialog.h Cloud/TodaysPlanBodyMeasures.h \
            Cloud/WithingsDownload.h Cloud/Strava.h Cloud/CyclingAnalytics.h Cloud/RideWithGPS.h \
            Cloud/TrainingsTageBuch.h Cloud/Selfloops.h Cloud/Velohero.h Cloud/SportsPlusHealth.h \
-           Cloud/AddCloudWizard.h Cloud/Withings.h Cloud/MeasuresDownload.h Cloud/Xert.h
+           Cloud/AddCloudWizard.h Cloud/Withings.h Cloud/MeasuresDownload.h Cloud/Xert.h \
```
Comment (amtriathlon) [inline src/src.pro:708]: No need for this change, please remove it.

File: src/src.pro [hunk 9]
```diff
@@ -807,7 +807,7 @@ SOURCES += Cloud/CalendarDownload.cpp Cloud/CloudService.cpp \
            Cloud/LocalFileStore.cpp Cloud/OAuthDialog.cpp Cloud/TodaysPlanBodyMeasures.cpp \
            Cloud/WithingsDownload.cpp Cloud/Strava.cpp Cloud/CyclingAnalytics.cpp Cloud/RideWithGPS.cpp \
            Cloud/TrainingsTageBuch.cpp Cloud/Selfloops.cpp Cloud/Velohero.cpp Cloud/SportsPlusHealth.cpp \
-           Cloud/AddCloudWizard.cpp Cloud/Withings.cpp Cloud/MeasuresDownload.cpp Cloud/Xert.cpp
+           Cloud/AddCloudWizard.cpp Cloud/Withings.cpp Cloud/MeasuresDownload.cpp Cloud/Xert.cpp \
```
Comment (amtriathlon) [inline src/src.pro:810]: No need for this change, please remove it.

File: src/Cloud/OAuthDialog.cpp [hunk 10]
```diff
@@ -467,6 +468,7 @@ OAuthDialog::networkRequestFinished(QNetworkReply *reply)
             refresh_token = document.object()["refresh_token"].toString();
             access_token = document.object()["access_token"].toString();
             if (site == POLAR)  polar_userid = document.object()["x_user_id"].toDouble();
+            //if (site == POLAR)  transaction_id = document.object()["transaction-id"].toString();
```
Comment (amtriathlon) [inline src/Cloud/OAuthDialog.cpp:471]: Please don't add commented out code

File: src/Cloud/OAuthDialog.cpp [hunk 11]
```diff
@@ -509,6 +511,9 @@ OAuthDialog::networkRequestFinished(QNetworkReply *reply)
 
             service->setSetting(GC_POLARFLOW_TOKEN, access_token);
             service->setSetting(GC_POLARFLOW_USER_ID, polar_userid);
+            service->setSetting(GC_POLARFLOW_TRANSACTION_ID, transaction_id);
+            //service->setSetting(GC_POLARFLOW_LAST_REFRESH, QDateTime::currentDateTime());
```
Comment (amtriathlon) [inline src/Cloud/OAuthDialog.cpp:515]: Please don't add commented out code

**Comment by amtriathlon** [comment]
Build is failing due to the use of and instead of && logical operator.

**Comment by ssj2son-gohan-ne** [comment]
> Build is failing due to the use of and instead of && logical operator.
Thanks (:

think i did all the changes you advised...

**Comment by amtriathlon** [comment]
Since "Detailed Sport Info" field seems to have the same purpose as current "SubSport" field, I would avoid to create a new field and use the existing one.

**Comment by ssj2son-gohan-ne** [comment]
> Since "Detailed Sport Info" field seems to have the same purpose as current "SubSport" field, I would avoid to create a new field and use the existing one.

i use all the fields with separate info's for filtering, but if insisted i can change to fill in "Subsport"
Screenshot from 2022-05-07 08-34-55

Besides that i don't know where to put the call for the "PolarFlow::Close()"  (Commit Transaction).
If done around this line [https://github.com/GoldenCheetah/GoldenCheetah/pull/4145/files#diff-a5e3571c9c2a895f2d9454e078db2e4a2ef2193b4838a271d9d49243b1dd2784L163](url) it is too early if done before and after this line it is not read...

**Comment by amtriathlon** [comment]
> > Since "Detailed Sport Info" field seems to have the same purpose as current "SubSport" field, I would avoid to create a new field and use the existing one.
> 
> i use all the fields with separate info's for filtering, but if insisted i can change to fill in "Subsport"

Yes, please do. SubSport is already available for that purpose, there is no point to create a duplicate.

> Besides that i don't know where to put the call for the "PolarFlow::Close()" (Commit Transaction).

Nowhere, close service entry point is called by GC as explained in https://github.com/GoldenCheetah/GoldenCheetah/wiki/Using-the-Cloud-Services-API#void-close

>If done around this line it is too early if done before and after this line it is not read...
> 
> https://github.com/GoldenCheetah/GoldenCheetah/pull/4145/files#diff-a5e3571c9c2a895f2d9454e078db2e4a2ef2193b4838a271d9d49243b1dd2784L163

Don’t you understand what return do?

**Comment by ssj2son-gohan-ne** [comment]
Sorry for the long time that has passed.
This years Highlight was last weeks Ironman Frankfurt and so there was not much time for programming and further progress in the last 6 weeks or 8 weeks because of training.
 
> Yes, please do. SubSport is already available for that purpose, there is no point to create a duplicate
Now Polar Flow Import writes into SubSport Field.

> Nowhere, close service entry point is called by GC as explained in https://github.com/GoldenCheetah/GoldenCheetah/wiki/Using-the-Cloud-Services-API#void-close
> 
So i thought, i unterstood ...
But close is not called, this i do not understand.

> Don’t you understand what return does?
It seem so, are you willing to explain it to me?
Inspected Strava, and other Online-Services, but did not find where the difference is...

Sorry for the long time passed by.

**Comment by ssj2son-gohan-ne** [comment]
Now the transaction is committed after retrieving last activity / exercise and when there is no new content.
